src/inference/graph_inference.cpp

```diff
@@ -920,60 +933,94 @@ class GraphNodeAssignment : public Nice2Assignment {
           factors_candidates.push_back(factors[j]);
         }
       }
+
+      // Go over all factors containing the labels in the "given" set, and create
+      // the initial permutation. It includes all labels in a factor that are not in the "given" set.
+      Factor giv_labels_copy = giv_labels;
       for (size_t j = 0; j < factors_candidates.size(); ++j) {
-        Factor current_candidate = factors_candidates[j];
-        std::set<int> distinct_labels;
         std::vector<int> candidate_inf_labels;
-        candidate_inf_labels.reserve(current_candidate.size());
-        std::unordered_set<std::vector<int>> permutations;
-        for (auto label = current_candidate.begin(); label != current_candidate.end(); ++label) {
-          if (giv_labels.count(*label) == 0) {
+        candidate_inf_labels.reserve(factors_candidates[j].size());
+        for (auto label = factors_candidates[j].begin(); label != factors_candidates[j].end(); ++label) {
+          // Check if the label belongs to the "given" set.
+          if (giv_labels.count(*label) > 0) {
```

Comment:
This is not a good style. You do two lookups in the map. First, you call count and then you do find. 2 lookups => slow. What you can do is to do only find and then to know if you found anything, just check if it != giv_labels.end()